File: src/main/java/com/googongill/aditory/domain/Category.java
package com.googongill.aditory.domain;

import com.googongill.aditory.domain.enums.CategoryState;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Category extends BaseEntity {
    @Id @Column(name = "category_id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String categoryName;
    private String asCategoryName;
    @Enumerated(EnumType.STRING)
    private CategoryState categoryState;
    private Integer viewCount;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @OneToMany(mappedBy = "category", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Link> links = new ArrayList<>();

    @OneToMany(mappedBy = "category", orphanRemoval = true)
    private List<CategoryLike> categoryLikes = new ArrayList<>();

    public Category(String categoryName) {
        this.categoryName = categoryName;
    }

    public Category(String categoryName, String asCategoryName, User user) {
        this.categoryName = categoryName;
        this.asCategoryName = asCategoryName;
        this.viewCount = 0;
        this.categoryState = CategoryState.PRIVATE;
        this.user = user;
    }

    // 연관관계 메서드
    public void setUser(User user) {
        this.user = user;
    }

    public void addLink(Link link) {
        this.links.add(link);
        link.setCategory(this);
    }

    public void addCategoryLike(CategoryLike categoryLike) {
        this.categoryLikes.add(categoryLike);
        categoryLike.setCategory(this);
    }

    public void deleteCategoryLike(CategoryLike categoryLike) {
        this.categoryLikes.remove(categoryLike);
    }

    public void updateCategoryInfo(String categoryName, String asCategoryName, CategoryState categoryState) {
        this.categoryName = categoryName;
        this.asCategoryName = asCategoryName;
        this.categoryState = categoryState;
    }
}

File: src/main/java/com/googongill/aditory/domain/Category.kt
package com.googongill.aditory.domain

import com.googongill.aditory.domain.enums.CategoryState
import jakarta.persistence.*

@Entity
class Category : BaseEntity {
    @Id
    @Column(name = "category_id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    var categoryName: String
    lateinit var asCategoryName: String

    @Enumerated(EnumType.STRING)
    lateinit var categoryState: CategoryState

    var viewCount: Int? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    lateinit var user: User

    @OneToMany(mappedBy = "category", cascade = [CascadeType.ALL], orphanRemoval = true)
    var links: MutableList<Link> = ArrayList()

    @OneToMany(mappedBy = "category", orphanRemoval = true)
    var categoryLikes: MutableList<CategoryLike> = ArrayList()

    constructor(categoryName: String) {
        this.categoryName = categoryName
    }

    constructor(categoryName: String, asCategoryName: String, user: User) {
        this.categoryName = categoryName
        this.asCategoryName = asCategoryName
        viewCount = 0
        categoryState = CategoryState.PRIVATE
        this.user = user
    }

    // 연관관계 메서드
    fun createUser(user: User) {
        this.user = user
    }

    fun addLink(link: Link) {
        links.add(link)
        link.updateCategory(this)
    }

    fun addCategoryLike(categoryLike: CategoryLike) {
        categoryLikes.add(categoryLike)
        categoryLike.updateCategory(this)
    }

    fun deleteCategoryLike(categoryLike: CategoryLike) {
        categoryLikes.remove(categoryLike)
    }

    fun updateCategoryInfo(categoryName: String, asCategoryName: String, categoryState: CategoryState) {
        this.categoryName = categoryName
        this.asCategoryName = asCategoryName
        this.categoryState = categoryState
    }
}
